### PR TITLE
EASY-1325: search servlet

### PR DESCRIPTION
**Pull request by jo-pol**
Fixes EASY-1325

#### When applied 
* the solr core won't produce `attr_` prefixes in indexed field names, though now again images and pakbon contents are not indexed, just their metadata (#8 embedded intis PR)
* filters queries on accessibility and data available
* apply optional basic authentication
* [x] filters on accessible-to unless the (optionally) authenticated user is the owner of the dataset or the user is an archivist or administrator

Next PR's
* rename the project to easy-file-index (and change easy-build accordingly)
* new issues to distill from the todo's in the source

#### How should this be manually tested?

```
mvn clean install && vagrant destroy -f && vagrant up
curl -X POST 'http://test.dans.knaw.nl:20150/fileindex/init'
curl -v -u user001:user001 'http://test.dans.knaw.nl:20150/filesearch?text=voor' | grep file_path
```

The `-v` option shows the response headers, the `-u` option is also optional.
`user001` and `easyadmin` will get two files, other users just one when searching for the word `voor`.
See the new readme for variations of both curl commands and command line variants of the first curl command.

The bag store on the local VM is initialized with two test bags which get indexed with the first curl command. It takes a while but you can load additional bags by copying the content of `/tmp/pdbs.tar` on `teasy.dans.knaw.nl` into `/data/bagstores/pdbs` of the local VM.

**Note / prerequisites for vagrant up**
I had the following directories avaiable in `~/.ansible/roles/`, synchronizing and running dtap might help

dans.capture-deployer-email
dans.easy-add-new-license
dans.easy-bag-index
dans.easy-bag-store
dans.easy-includes
dans.easy-ldap-dir
dans.easy-send-metadata-to-datacite
dans.easy-test-users
dans.local-test-vm-base
dans.local-yum-repo
dans.postgresql
dans.solr

#### Related pull requests on github
repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-update-solr4files-index    | #8     | accepted but ~~not yet merged~~ closed PR is included in this one [differences since then](https://github.com/DANS-KNAW-jp/easy-update-solr4files-index/pull/1/)

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.solr4files/SearchServlet.scala [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  // lucene seems not safe: Support for using any type of query parser as a nested clause
+  // edismax: supports the full Lucene query parser syntax, so possibly not safe too
+  // dismax: more like [...] Google [...] makes [...] appropriate [...] for many consumer applications
+  private val queryParser = "dismax"
+
+  get("/") {
+    params.get("text")
+      .map(q => respond(app.search(createQuery(q))))
+      .getOrElse(BadRequest("filesearch requires param 'text' (a solr dismax query), got " + params.mkString("[",",","]")))
+  }
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.solr4files/SearchServlet.scala]: Did not implement a command line version. In the future we might want to support many options of the `SolrQuery`. Mapping them to HTML paths and queries alone will be complex/_elaborate_ enough, perhaps subject to a lot of debate.

**Comment by jo-pol** [comment]
This PR also includes the changes of the not yet merged https://github.com/DANS-KNAW/easy-update-solr4files-index/pull/8

**Comment by jo-pol** [comment]
An experiment: a local PR for the differences between the not yet merged https://github.com/DANS-KNAW/easy-update-solr4files-index/pull/8 and https://github.com/DANS-KNAW/easy-update-solr4files-index/pull/9

File: src/main/scala/nl.knaw.dans.easy.solr4files/SearchServlet.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  private def createQuery(query: String) = {
+    new SolrQuery() {
+      setQuery(query)
+      addFilterQuery("easy_file_accessible_to:ANONYMOUS + easy_file_accessible_to:KNOWN")
+      // TODO filter query on available date
+      setFields("easy_dataset_*", "easy_file_*")
+      setStart(0)
+      setRows(10)
+      setTimeAllowed(5000) // 5 seconds TODO make configurable
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.solr4files/SearchServlet.scala]: why make this configurable? I'd say 5 seconds is already a big interval to wait for these first 10 results

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala [hunk 3]
```diff
@@ -54,22 +59,61 @@ trait Solr extends DebugEnhancedLogging {
...
-      .flatMap(checkUpdateResponseStatus)
+  def deleteDocuments(query: String): Try[UpdateResponse] = {
+    val q = new SolrQuery {
+      set("q", query)
+    }
+    Try(solrClient.deleteByQuery(q.getQuery))
+      .flatMap(checkResponseStatus)
       .recoverWith {
-        case t: HttpSolrClient.RemoteSolrException if t.getRootThrowable.endsWith("ParseException") =>
+        case t: HttpSolrClient.RemoteSolrException if isPareException(t) =>
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala]: `isParseException`

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala [hunk 4]
```diff
@@ -54,22 +59,61 @@ trait Solr extends DebugEnhancedLogging {
...
-        case t: SolrServerException =>
-          Failure(SolrDeleteException(query, t))
         case t =>
           Failure(SolrDeleteException(query, t))
       }
   }
 
-  def commit(): Try[Unit] = {
+  private def toJson(solrDocumentList: SolrDocumentList, query: SolrQuery): Try[String] = {
+    def toScalaMap(fieldValueMap: util.Map[String, AnyRef]) = {
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala]: why not make this a proper extension method? Especially if you want to mimic `asScala` I think that is preferable.

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala [hunk 5]
```diff
@@ -54,22 +59,61 @@ trait Solr extends DebugEnhancedLogging {
...
+      // asScala is not good enough, serialization needs scala case classes
+      // less verbose than TypeHints, after all we need only one direction
+      fieldValueMap.keySet().asScala.map(key => key -> fieldValueMap.get(key))
+    }
+
+    val found = solrDocumentList.getNumFound
+    Try(write(Map("header" -> Map(
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala]: is this whole thing supposed to represent JSON? If so, have you tried using the library's syntax instead. See also https://github.com/DANS-KNAW/easy-stage-dataset/blob/master/src/main/scala/nl.knaw.dans.easy.stage/lib/JSON.scala

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala [hunk 6]
```diff
@@ -54,22 +59,61 @@ trait Solr extends DebugEnhancedLogging {
...
+  def search(query: SolrQuery): Try[String] = {
+    (for {
+      response <- Try(solrClient.query(query))
+      _ <- checkResponseStatus(response)
+      json <- toJson(response.getResults, query)
+    } yield json).recoverWith {
+      case t: HttpSolrClient.RemoteSolrException if isPareException(t) =>
+        Failure(SolrBadRequestException(t.getMessage, t))
+      case t =>
+        t.printStackTrace()
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala]: I assume this line is going to be deleted sometime soon?

File: src/test/scala/nl.knaw.dans.easy.solr4files/SearchServletSpec.scala [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  addServlet(new SearchServlet(app), "/*")
+
+  "get /" should "complain about missing argument" in {
+    get("/?q=something") {
+      body shouldBe "filesearch requires param 'text' (a solr dismax query), got [q -> something]"
+      status shouldBe SC_BAD_REQUEST
+    }
+  }
+
+  it should "return json" in {
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.solr4files/SearchServletSpec.scala:74]: wouldn't this be better understandable/cleaner/clearer if you parsed your response back to JSON and then compare it to the expected JSON structure?

File: src/test/scala/nl.knaw.dans.easy.solr4files/UpdateServletSpec.scala [hunk 8]
```diff
@@ -145,7 +145,7 @@ class UpdateServletSpec extends TestSupportFixture
 
   it should "complain about the required query" in {
     delete("/") {
-      body shouldBe "delete requires param 'q': a solr query"
+      body shouldBe "delete requires param 'q': got []"
```
Comment (rvanheest) [inline src/test/scala/nl.knaw.dans.easy.solr4files/UpdateServletSpec.scala]: Is this the actual message that users get to see? I don't like the `[]`. You better return something useful here!

**Review by rvanheest — CHANGES_REQUESTED** [review]

File: src/main/scala/nl.knaw.dans.easy.solr4files/SearchServlet.scala (hunk 2 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.solr4files/SearchServlet.scala]: start and rows are intended as configurable by the user with additional http arguments. TimeAllowed and a maximum value for rows are intended to become configurable through application properties.

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala (hunk 4 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala]: Will work that out when implementing data-available, I might get second thoughts about mapping versus TypeHints.

File: src/main/scala/nl.knaw.dans.easy.solr4files/EasyUpdateSolr4filesIndexService.scala [hunk 9]
```diff
@@ -33,7 +33,8 @@ class EasyUpdateSolr4filesIndexService(val serverPort: Int, app: EasyUpdateSolr4
         override def probeForCycleClass(classLoader: ClassLoader): (String, LifeCycle) = {
           ("Solr4files-lifecycle", new LifeCycle {
             override def init(context: ServletContext): Unit = {
-              context.mount(new EasyUpdateSolr4filesIndexServlet(app), "/fileindex")
+              context.mount(new UpdateServlet(app), "/fileindex")
+              context.mount(new SearchServlet(app), "/filesearch")
```
Comment (PaulBoon) [inline src/main/scala/nl.knaw.dans.easy.solr4files/EasyUpdateSolr4filesIndexService.scala:37]: My plan for the search API was 'easy.dans.knaw.nl/fileitem/search', so have a 'fileitem' resource with a search on it. 
Maybe we have Jan and Linda give their opinions, but it seems easy to change at this stage in the development. 
As soon as we have the API in production, we are kind of stuck with it.

Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.solr4files/EasyUpdateSolr4filesIndexService.scala:37]: The proxies should map `http://deasy.dans.knaw.nl/files/search` to `http://deasy.dans.knaw.nl:20150/filesearch` (and `http://deasy.dans.knaw.nl/files/download` to another servlet)

File: src/main/scala/nl.knaw.dans.easy.solr4files/SearchServlet.scala (hunk 2 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.solr4files/SearchServlet.scala]: @rvanheest I tried:

      setRows(params
        .get("limit")
        .withFilter(_.matches("[1-9][0-9]*"))
        .map(_.toInt)
        .getOrElse(10)
      )

what resulted in:

```
Error:(67, 19) type mismatch;
 found   : Any
 required: Integer
        .getOrElse(10)
```

any clue? Perhaps there is something easier.

File: src/test/scala/nl.knaw.dans.easy.solr4files/SearchServletSpec.scala (hunk 7 above)
Comment (jo-pol) [reply src/test/scala/nl.knaw.dans.easy.solr4files/SearchServletSpec.scala:74]: The message applied to https://github.com/DANS-KNAW-jp/easy-update-solr4files-index/blob/b54b5b3a9a731bdd567d6dadf1a884ea649579d5/src/test/scala/nl.knaw.dans.easy.solr4files/SearchServletSpec.scala#L77-L92 these lines are changed, the header now has a fixed order simplifying the test.

File: src/main/ansible/vagrant.yml [hunk 10]
```diff
@@ -119,18 +130,47 @@
...
+- hosts: "test"
+  vars:
+    shared_ldap_admin_user: "cn=ldapadmin,dc=dans,dc=knaw,dc=nl"
+    shared_ldap_url: "ldap://localhost"
+    easy_ldap_dir_dans_schema_location: "/etc/openldap/slapd.d/cn=config/cn=schema/cn={12}dans.ldif"
+    easy_ldap_dir_easy_schema_location: "/etc/openldap/slapd.d/cn=config/cn=schema/cn={13}easy.ldif"
+    easy_ldap_dir_easyadmin_email: "noreply@whatever.dans.knaw.nl"
+    easy_ldap_dir_easyadmin_password: "easyadmin"
+    easy_ldap_dir_ldapadmin_password: "ldapadmin"
+    easy_test_users_ldapadmin_password: "ldapadmin"
```
Comment (jo-pol) [inline src/main/ansible/vagrant.yml]: @janvanmansum 
Don't understand much of it, just grabbed something from different places. Do I need all the variables? ApacheDirectoryStudio does show the users with the above email address for easyadmin, so it isn't talking to some other instance.

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/LdapAuthenticationComponent.scala [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  val usersParentEntry: String
+  val securityPrincipal: String
+  val securityCredentials: String
+  val ldapProviderUrl: URI
+  val initialContextFactory: String = "com.sun.jndi.ldap.LdapCtxFactory"
+
+  trait LdapAuthentication extends Authentication {
+
+    def getUser(userName: String, password: String): Try[User] = Try {
+      val ctxt = new InitialLdapContext(ldapEnv(userName, password), null)
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.solr4files/components/LdapAuthenticationComponent.scala]: This doesn't work. Suppose I need two stage somehow. First establish a connection with the principal configuration parameters, then verify username and password.

File: src/main/ansible/vagrant.yml [hunk 12]
```diff
@@ -60,14 +60,22 @@
...
+  - name: Copy space-travel bag to server
+    copy: # bag-store knows it with uuid 40594b6d-8378-4260-b96b-13b57beadf7c
       src: "../../test/resources/vault/stores/pdbs/bags/9da0541a-d2c8-432e-8129-979a9830b427"
       dest: "{{ stores_dir }}/pdbs/ab/123456789012345678901234567890"
       mode: "0775"
       directory_mode: "0775"
     when: "bag_dir_stat.stat.exists == False"
 
+  - name: Copy test-bag to server
+    copy: # bag-store knows it with uuid ab123456-7890-1234-5678-901234567890
```
Comment (PaulBoon) [inline src/main/ansible/vagrant.yml:72]: Looks like the comment is swapped with the one above where the dest dir starts with 'ab'

Comment (jo-pol) [reply src/main/ansible/vagrant.yml:72]: No, it is just a coincidence. I once thought the folder names (after `ab`/ respective `40/`) where the UUID's of the bags. They appeared to be unrelated. I discovered the actual values after indexing the bags in solar. As a convenience for manual tests I added the actual UUID's of the bags as comment.

File: src/main/scala/nl.knaw.dans.easy.solr4files/SearchServlet.scala [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  // edismax: supports the full Lucene query parser syntax, so possibly not safe too
+  // dismax: more like [...] Google [...] makes [...] appropriate [...] for many consumer applications
+  private val queryParser = "dismax"
+
+  get("/") {
+    // no command line equivalent, use http://localhost:8983/solr/#/fileitems/query
+    // or for example:           curl 'http://localhost:8983/solr/fileitems/query?q=*'
+    (params.get("text"), app.authenticate(new BasicAuthRequest(request))) match {
+      case (None, _) => BadRequest("filesearch requires param 'text' (a solr dismax query), got " + params.asString)
+      case (Some(q), Success(user)) => respond(app.search(createQuery(q, user)))
```
Comment (PaulBoon) [inline src/main/scala/nl.knaw.dans.easy.solr4files/SearchServlet.scala]: Where is the unauthenticated request handled?

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/AuthenticationComponent.scala [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+import scala.util.{ Success, Try }
+
+trait AuthenticationComponent extends DebugEnhancedLogging {
+
+  val authentication: Authentication
+
+  trait Authentication {
+    def authenticate(authRequest: BasicAuthRequest): Try[Option[User]] = {
+
+      if (authRequest.providesAuth && authRequest.isBasicAuth) {
```
Comment (PaulBoon) [inline src/main/scala/nl.knaw.dans.easy.solr4files/components/AuthenticationComponent.scala]: If it provides authentication, but it's not basic, maybe we should give a message about this unsupported authentication method?
Probably give 'unauthorised' at least the caller (client) then knows that the intended request failed.  
For the time being we could leave it out, also I don't know how we could test it.

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala [hunk 15]
```diff
@@ -54,22 +61,59 @@ trait Solr extends DebugEnhancedLogging {
...
+    val result =
+      ("header" -> (
+        ("text" -> query.getQuery) ~
+          ("skip" -> query.getStart.toInt) ~
+          ("limit" -> query.getRows.toInt) ~
+          ("time_allowed" -> query.getTimeAllowed.toInt) ~
+          ("found" -> numFound) ~
+          ("returned" -> fileItems.size)
+        )) ~
+        ("fileitems" -> fileItems)
```
Comment (PaulBoon) [inline src/main/scala/nl.knaw.dans.easy.solr4files/components/Solr.scala:93]: Looks like we are passing on the fields we get from Solr response, but possibly we need to filter this (by Solr request) and even rename fieldnames. But for now it's ok.  
Maybe put TODO here?

**Review by PaulBoon — APPROVED** [review]

File: src/main/scala/nl.knaw.dans.easy.solr4files/components/AuthenticationComponent.scala (hunk 14 above)
Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.solr4files/components/AuthenticationComponent.scala]: I guess I have to add header fields to get requests in SearchServletSpec. A curl example:

```
curl -v -u user002:user002 'http://test.dans.knaw.nl:20150/filesearch?text=voor' | grep file_path
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 192.168.33.32...
* TCP_NODELAY set
* Connected to test.dans.knaw.nl (192.168.33.32) port 20150 (#0)
* Server auth using Basic with user 'user002'
> GET /filesearch?text=voor HTTP/1.1
> Host: test.dans.knaw.nl:20150
> Authorization: Basic dXNlcjAwMjp1c2VyMDAy
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: text/plain;charset=UTF-8
< Content-Length: 144
< Server: Jetty(8.2.0.v20160908)
<
{ [144 bytes data]
100   144  100   144    0     0   5507      0 --:--:-- --:--:-- --:--:--  5538
* Connection #0 to host test.dans.knaw.nl left intact
```

**Pull request merged**